scene_synchronizer.cpp

```diff
@@ -3239,196 +3210,259 @@ void ClientSynchronizer::process_controllers_recovery(real_t p_delta) {
+			// mechanism, may have insert a no sync node.
+			// The check is here because I feel it more clear, here.
+			continue;
+		}
 
-			// Step 3 -- Process the Actions.
 #ifdef DEBUG_ENABLED
-			// This can't happen because the client stores a snapshot for each frame.
-			CRASH_COND(client_snapshots[i].input_id == checkable_input_id + i);
+		// The parser make sure to properly initialize the snapshot variable
+		// array size. So the following condition is always `false`.
+		CRASH_COND(uint32_t(server_snapshots.front().node_vars[p_nodes_to_recover[i]->id].size()) != p_nodes_to_recover[i]->vars.size());
 #endif
-			for (int a_i = 0; a_i < client_snapshots[i].actions.size(); a_i += 1) {
-				// Leave me alone, I don't want to make `execute()` const. 😫
-				NetActionProcessor(client_snapshots[i].actions[a_i].processor).execute();
-				SceneSynchronizerDebugger::singleton()->debug_print(scene_synchronizer, "Rewind, processed Action: " + String(client_snapshots[i].actions[a_i].processor));
-			}
-
-			// Step 4 -- Pull node changes and Update snapshots.
-			for (uint32_t r = 0; r < nodes_to_recover.size(); r += 1) {
-				if (nodes_to_recover[r]->sync_enabled == false) {
-					// This node is not sync.
-					continue;
-				}
-				// Pull changes
-				scene_synchronizer->pull_node_changes(nodes_to_recover[r]);
 
-				// Update client snapshot.
-				if (uint32_t(client_snapshots[i].node_vars.size()) <= nodes_to_recover[r]->id) {
-					client_snapshots[i].node_vars.resize(nodes_to_recover[r]->id + 1);
-				}
+		Node *node = p_nodes_to_recover[i]->node;
+		const Vector<NetUtility::Var> s_vars = server_snapshots.front().node_vars[p_nodes_to_recover[i]->id];
+		const NetUtility::Var *s_vars_ptr = s_vars.ptr();
 
-				Vector<NetUtility::Var> *snap_node_vars = client_snapshots[i].node_vars.ptrw() + nodes_to_recover[r]->id;
-				snap_node_vars->resize(nodes_to_recover[r]->vars.size());
+		SceneSynchronizerDebugger::singleton()->debug_print(scene_synchronizer, "Full reset node: " + node->get_path());
 
-				NetUtility::Var *vars = snap_node_vars->ptrw();
-				for (uint32_t v = 0; v < nodes_to_recover[r]->vars.size(); v += 1) {
-					vars[v] = nodes_to_recover[r]->vars[v].var;
-				}
+		for (int v = 0; v < s_vars.size(); v += 1) {
+			if (s_vars_ptr[v].name == StringName()) {
+				// This variable was not set, skip it.
+				continue;
 			}
-			scene_synchronizer->change_events_flush();
+
+			const Variant current_val = p_nodes_to_recover[i]->vars[v].var.value;
+			p_nodes_to_recover[i]->vars[v].var.value = s_vars_ptr[v].value.duplicate(true);
+			node->set(s_vars_ptr[v].name, s_vars_ptr[v].value);
+
+			SceneSynchronizerDebugger::singleton()->debug_print(scene_synchronizer, " |- Variable: " + s_vars_ptr[v].name + " New value: " + s_vars_ptr[v].value);
+			scene_synchronizer->change_event_add(
+					p_nodes_to_recover[i],
+					v,
+					current_val);
 		}
+	}
+	scene_synchronizer->change_events_flush();
+}
+
+void ClientSynchronizer::__pcr__rewind(
```

Comment:
I know this says rewind, but this is the part of the code that replays the client input buffer, yeah?

Reply:
Yes, into this function you can find the following lines:
```
[3409]       p_nodes_to_recover[r]->process(p_delta);
```

```
[3422]       p_controller->process_instant(i, p_delta)
```


```
[3433]       NetActionProcessor(client_snapshots[i].actions[a_i].processor).execute();
```

and so on.